### ISSUE TITLE
rails credentials not working for environments / secret_key_base

### ISSUE THREAD

**Issue by chmich** [rails/rails#46636]
### Steps to reproduce
`rails credentials:edit` works well on development.
by example, having a line in it like `user_password: "xxxx"` is accessible by `Rails.application.credentials[:user_password]`

and, i could deploy my new app to the production server, on environment `staging`

then, i set the same `user_password`, on credentials, on staging by `rails credentials:edit --environment=staging` 
deploy didnt work because its missing `secret_key_base`

so far, so clear.

regarding to [Railsguides/configuring](https://guides.rubyonrails.org/configuring.html#config-secret-key-base) and [api](https://api.rubyonrails.org/v7.0.4/classes/Rails/Application.html#method-i-secret_key_base) two ways should work: 

- set a environment variable on linux server `SECRET_KEY_BASE`
- set a line like `secret_key_base: xxx` by `rails credentials:edit --environment=staging` 

should solve this but didnt

i had to set explicitly a config like `config.secret_key_base = ENV["SECRET_KEY_BASE"]` inside `config/environments/` for let the rails pickup the keybase from linux-env.

and, i am not able to use the credentials for other purposes, like above mentioned `user_password`. On development that all works fine and as long i have only one secret and never had triggered environment specific credentials, rails picks up all fine from the development environment.

Rails-7.0.4
Ruby-3.1.2

**Comment by skipkayhil** [comment]
Hey @chmich, thanks for opening an issue! Can you create a sample app in a public repo with explicit steps to demonstrate this issue? Without this it is hard to know exactly what could be going wrong

**Comment by chmich** [comment]
Thanks @skipkayhil for the quick response!

you should be able to reproduce this just by `rails new ...` and following the steps mentioned above in this issue.

But, while writing this i found something new for me:

If created environment specific credentials by `rails credentials:edit --environment=staging` rails creates the folder `config/credentials/` with regarding files. **If deleting this folder rails is using the firstly created credentials** (development..?) for all environments. So, i did this and my `secret_key_base` works from credentials on staging and also the others, custom credentials i defined in the one file are working also for staging!

So, my questions are: 
Why using this environment specific credentials?
The are cumbersome and confusing!
Do they bring a security advantage? (there are no environment specific encryption-keys but the one and only master-key)
Why the where introduced?

If i need environment specific credentials i can make:

```
for_all_environments: "abc"
development: 
   development_specific_cred: "cde"
```

and so on.

Why not kill this environment specific credentials instead of reparing it?
Is there anyone who needs it?

In other words:
similar to [41830](https://github.com/rails/rails/issues/41830) my biggest problem are the rails-docs because all this behaviour is not described.

My problem is not that environment-specific credentials are not working, my problem was that i meant they are  the way-to-go. I am more happier with one credentials file for all envs.

So, my problem is solved.
I will continue no-environment-specific credentials but i am courious of this questions.

Thanks,
Christian

**Comment by psantos10** [comment]
When you run `rails credentials:edit --environment=staging` , `config/credentials/staging. Key` file is generated. 

Make sure that the value on file is the one you use as `RAILS_MASTER_KEY` on staging environment

**Comment by chmich** [comment]
thanks, @psantos10 for answer!

But, regarding to [docs](https://guides.rubyonrails.org/security.html#custom-credentials), `ENV["RAILS_MASTER_KEY"]` should contain the `config/master.key` from development machine, on the production / staging machine. I have the master key on `shared/config/master.key` on the production machine and this works.

This sustains my assumption that the environment specific credential files are bringing no advantage on security, because: The person which haves the master key can always encrypt all credential environments.

**Comment by psantos10** [comment]
`RAILS_MASTER_KEY` should contain `config/master. Key` if you are using `config/credentials.yml.enc`. But that is not anymore your. Since you are ysung environment credentials, the RAILS_MASTER_KEY must be the one specific for that environment and not anymore `config/master.key`.

**Comment by chmich** [comment]
Thanks, @psantos10 for this explanation! i`ll test this! This makes sense. where is this to find at the docs?

**Comment by psantos10** [comment]
> rails credentials:edit --environment

Check the docs here: https://api.rubyonrails.org/classes/Rails/Application.html#method-i-credentials

**Comment by chmich** [comment]
Thanks @psantos10, now my credentials are working for staging and that makes sense and is a advantage on security. I killed the `config/master.key` and use config/credentials/staging.key` (the same is included in .gitignore).

Sorry, i have't read the  api.rubyonrails, it's explaining it well.

Cool, thanks!
